core/solver/batch_bicgstab.cpp

```diff
@@ -78,105 +75,42 @@ std::unique_ptr<BatchLinOp> BatchBicgstab<ValueType>::conj_transpose() const
-    const Mtx* a_scaled{};
-    const Vector* b_scaled{};
-    auto a_scaled_smart = Mtx::create(exec);
-    auto b_scaled_smart = Vector::create(exec);
-    if (to_scale && !acsr) {
-        GKO_NOT_SUPPORTED(system_matrix_);
-    }
-
-    // copies to scale
-    if (to_scale) {
-        a_scaled_smart->copy_from(acsr);
-        b_scaled_smart->copy_from(dense_b);
-        exec->run(batch_bicgstab::make_pre_diag_scale_system(
-            this->get_left_scaling_vector(), this->get_right_scaling_vector(),
-            a_scaled_smart.get(), b_scaled_smart.get()));
-        a_scaled = a_scaled_smart.get();
-        b_scaled = b_scaled_smart.get();
-    } else if (acsr) {
-        a_scaled = acsr;
-        b_scaled = dense_b;
-    } else {
-        b_scaled = dense_b;
-    }
+    using Dense = matrix::BatchDense<ValueType>;
     const kernels::batch_bicgstab::BatchBicgstabOptions<
         remove_complex<ValueType>>
         opts{parameters_.preconditioner, parameters_.max_iterations,
              parameters_.residual_tol, parameters_.tolerance_type};
-
-    log::BatchLogData<ValueType> logdata;
-
-    // allocate logging arrays assuming uniform size batch
-    // GKO_ASSERT(dense_b->stores_equal_sizes());
-
-    const size_type num_rhs = dense_b->get_size().at(0)[1];
-    const size_type num_batches = dense_b->get_num_batch_entries();
-    batch_dim<> sizes(num_batches, dim<2>{1, num_rhs});
-
-    logdata.res_norms =
-        matrix::BatchDense<real_type>::create(this->get_executor(), sizes);
-    logdata.iter_counts.set_executor(this->get_executor());
-    logdata.iter_counts.resize_and_reset(num_rhs * num_batches);
-
-    if (acsr) {
-        exec->run(batch_bicgstab::make_apply(opts, a_scaled, b_scaled, dense_x,
-                                             logdata));
-    } else {
-        exec->run(batch_bicgstab::make_apply(opts, system_matrix_.get(),
-                                             b_scaled, dense_x, logdata));
-    }
-    this->template log<log::Logger::batch_solver_completed>(
-        logdata.iter_counts, logdata.res_norms.get());
-
-    if (to_scale) {
-        exec->run(batch_bicgstab::make_vec_scale(
-            this->get_right_scaling_vector(), dense_x));
-    }
-}
-
-
-template <typename ValueType>
-void BatchBicgstab<ValueType>::apply_impl(const BatchLinOp* alpha,
-                                          const BatchLinOp* b,
-                                          const BatchLinOp* beta,
-                                          BatchLinOp* x) const
-{
-    auto dense_x = as<matrix::BatchDense<ValueType>>(x);
-
-    auto x_clone = dense_x->clone();
-    this->apply(b, x_clone.get());
-    dense_x->scale(beta);
-    dense_x->add_scaled(alpha, x_clone.get());
+    auto exec = this->get_executor();
+    exec->run(batch_bicgstab::make_apply(
+        opts, mtx, as<const Dense>(b), as<Dense>(x),
+        *static_cast<log::BatchLogData<ValueType>*>(info.logdata)));
```

Comment:
If you would go with the inheritance version, this would look like:
```suggestion
        *as<log::BatchLogData<ValueType>*>(info)));
```